### ISSUE TITLE
Code Coverage for ms test v2 based tests in VS 2017 does not work.

### ISSUE THREAD

**Issue by RomanKernSW** [Microsoft/testfx#159]
## Description
> Please follow the link [click here](https://social.msdn.microsoft.com/Forums/de-DE/5b8640e7-1f9e-4040-9d98-90db8fb97c1a/build-agent-vs-2015-cannot-discover-mstest-v2-tests-continuous-integration-and-nightly-builds?forum=vsunittest)

## Steps to reproduce

> "C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\ID
E\CommonExtensions\Microsoft\TestWindow\vstest.console.exe" test.dll --> no tests found 
> "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise
\Common7\IDE\CommonExtensions\Microsoft\TestWindow\vstest.console.exe" test.dll --> tests found and executed

> runsettings are included and used [see here, almost the same](https://msdn.microsoft.com/en-us/library/jj159530.aspx)

## Expected behavior
> Visual Studio 2015 vstest.console should find all tests
> Visual Studio 2017 vstest.console should create code coverage

## Actual behavior
> No test found on VS 2015 version of vstest.console
> Test found and executed but no code coverage on VS 2017 

## Environment
> Build Agent is Window 10, with VS 2015. Or my Computer with Win 7, VS 2015 and 2017
> mstest v2 is 1.1.14 v
> Project build runtime .net 4.5.2

**Comment by AbhitejJohn** [comment]
@RomanKernSW : For VS 2015 when running locally, you need to provide a /testadapterpath pointing to bin\debug to get tests discovered. This logic is automatically built into the VS 2017 runner. I'm assuming this is a XAML build definition and not Build vnext. Since MSTest V2 is a nuget package based adapter please follow the steps [here](https://msdn.microsoft.com/en-us/library/ms253138%28v=vs.120%29.aspx#Enable a third-party unit framework) to get this working in the CI.

**Comment by RomanKernSW** [comment]
Thanks alot... I've been looking for this problem since last week, but couldnt find anything. Path to the custom testadapter solve the discovery problem!

**Comment by RomanKernSW** [comment]
I do have last question: is my runsettings config obsolete or why there is no code coverage shown if I run all in Visual Studio 2017. The 2015 version shows the code coverage

**Comment by AbhitejJohn** [comment]
@RomanKernSW : Can you share your runsettings please?

**Comment by RomanKernSW** [comment]
<?xml version="1.0" encoding="utf-8"?>


<RunSettings>
  <DataCollectionRunSettings>
    <DataCollectors>
      <DataCollector friendlyName="Code Coverage" uri="datacollector://Microsoft/CodeCoverage/2.0"
                     assemblyQualifiedName="Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector, Microsoft.VisualStudio.TraceCollector, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
        <Configuration>
          <CodeCoverage>
            

            
            <ModulePaths>
              <Exclude>
                <ModulePath>.*Test.dll</ModulePath>
                <ModulePath>.*Definition.dll</ModulePath>
                <ModulePath>.*FakeItEasy.dll</ModulePath>
                <ModulePath>.*FluentAssertions.dll</ModulePath>
                <ModulePath>.*FluentAssertions.core.dll</ModulePath>
                <ModulePath>.*Infrastructure.dll</ModulePath>
              </Exclude>
            </ModulePaths>

            
            <Attributes>
              <Exclude>

                <Attribute>^System\.Diagnostics\.DebuggerHiddenAttribute$</Attribute>
                <Attribute>^System\.Diagnostics\.DebuggerNonUserCodeAttribute$</Attribute>
                <Attribute>^System\.Runtime\.CompilerServices.CompilerGeneratedAttribute$</Attribute>
                <Attribute>^System\.CodeDom\.Compiler.GeneratedCodeAttribute$</Attribute>
                <Attribute>^System\.Diagnostics\.CodeAnalysis.ExcludeFromCodeCoverageAttribute$</Attribute>
              </Exclude>
            </Attributes>

            
            <CompanyNames>
              <Exclude>
                <CompanyName>.*microsoft.*</CompanyName>
              </Exclude>
            </CompanyNames>

            
            <PublicKeyTokens>
              
              <Exclude>
                <PublicKeyToken>^B77A5C561934E089$</PublicKeyToken>
                <PublicKeyToken>^B03F5F7F11D50A3A$</PublicKeyToken>
                <PublicKeyToken>^31BF3856AD364E35$</PublicKeyToken>
                <PublicKeyToken>^89845DCD8080CC91$</PublicKeyToken>
                <PublicKeyToken>^71E9BCE111E9429C$</PublicKeyToken>
                <PublicKeyToken>^8F50407C4E9E73B6$</PublicKeyToken>
                <PublicKeyToken>^E361AF139669C375$</PublicKeyToken>
              </Exclude>
            </PublicKeyTokens>

            
            <UseVerifiableInstrumentation>True</UseVerifiableInstrumentation>
            <AllowLowIntegrityProcesses>True</AllowLowIntegrityProcesses>
            <CollectFromChildProcesses>True</CollectFromChildProcesses>
            <CollectAspDotNet>False</CollectAspDotNet>

          </CodeCoverage>
        </Configuration>
      </DataCollector>
    </DataCollectors>
  </DataCollectionRunSettings>
</RunSettings>

**Comment by AbhitejJohn** [comment]
I do see the coverage file generated in `TestResults` folder at my end. One thing I had to fix however was an invalid character at line 56('<!—Don't forget "Attribute" at the end of the name -->  ') when I copied the content over. Removing that worked. It however does not show up the code coverage results window though. I'm not sure if you meant that.

**Comment by RomanKernSW** [comment]
OK, now it should be correct... couldnt put the xml here

&lt;?xml version=&quot;1.0&quot; encoding=&quot;utf-8&quot;?&gt;
&lt;RunSettings&gt;
  &lt;DataCollectionRunSettings&gt;
    &lt;DataCollectors&gt;
      &lt;DataCollector friendlyName=&quot;Code Coverage&quot; uri=&quot;datacollector://Microsoft/CodeCoverage/2.0&quot; assemblyQualifiedName=&quot;Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector, Microsoft.VisualStudio.TraceCollector, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a&quot;&gt;
        &lt;Configuration&gt;
          &lt;CodeCoverage&gt;
            &lt;ModulePaths&gt;
              &lt;Exclude&gt;
                &lt;ModulePath&gt;.*Test.dll&lt;/ModulePath&gt;
                &lt;ModulePath&gt;.*Definition.dll&lt;/ModulePath&gt;
                &lt;ModulePath&gt;.*FakeItEasy.dll&lt;/ModulePath&gt;
                &lt;ModulePath&gt;.*FluentAssertions.dll&lt;/ModulePath&gt;
                &lt;ModulePath&gt;.*FluentAssertions.core.dll&lt;/ModulePath&gt;
                &lt;ModulePath&gt;.*Infrastructure.dll&lt;/ModulePath&gt;
              &lt;/Exclude&gt;
            &lt;/ModulePaths&gt;
            &lt;Attributes&gt;
              &lt;Exclude&gt;
                &lt;Attribute&gt;^System\.Diagnostics\.DebuggerHiddenAttribute$&lt;/Attribute&gt;
                &lt;Attribute&gt;^System\.Diagnostics\.DebuggerNonUserCodeAttribute$&lt;/Attribute&gt;
                &lt;Attribute&gt;^System\.Runtime\.CompilerServices.CompilerGeneratedAttribute$&lt;/Attribute&gt;
                &lt;Attribute&gt;^System\.CodeDom\.Compiler.GeneratedCodeAttribute$&lt;/Attribute&gt;
                &lt;Attribute&gt;^System\.Diagnostics\.CodeAnalysis.ExcludeFromCodeCoverageAttribute$&lt;/Attribute&gt;
              &lt;/Exclude&gt;
            &lt;/Attributes&gt;
            &lt;CompanyNames&gt;
              &lt;Exclude&gt;
                &lt;CompanyName&gt;.*microsoft.*&lt;/CompanyName&gt;
              &lt;/Exclude&gt;
            &lt;/CompanyNames&gt;
            &lt;PublicKeyTokens&gt;
              &lt;Exclude&gt;
                &lt;PublicKeyToken&gt;^B77A5C561934E089$&lt;/PublicKeyToken&gt;
                &lt;PublicKeyToken&gt;^B03F5F7F11D50A3A$&lt;/PublicKeyToken&gt;
                &lt;PublicKeyToken&gt;^31BF3856AD364E35$&lt;/PublicKeyToken&gt;
                &lt;PublicKeyToken&gt;^89845DCD8080CC91$&lt;/PublicKeyToken&gt;
                &lt;PublicKeyToken&gt;^71E9BCE111E9429C$&lt;/PublicKeyToken&gt;
                &lt;PublicKeyToken&gt;^8F50407C4E9E73B6$&lt;/PublicKeyToken&gt;
                &lt;PublicKeyToken&gt;^E361AF139669C375$&lt;/PublicKeyToken&gt;
              &lt;/Exclude&gt;
            &lt;/PublicKeyTokens&gt;
            &lt;UseVerifiableInstrumentation&gt;True&lt;/UseVerifiableInstrumentation&gt;
            &lt;AllowLowIntegrityProcesses&gt;True&lt;/AllowLowIntegrityProcesses&gt;
            &lt;CollectFromChildProcesses&gt;True&lt;/CollectFromChildProcesses&gt;
            &lt;CollectAspDotNet&gt;False&lt;/CollectAspDotNet&gt;
          &lt;/CodeCoverage&gt;
        &lt;/Configuration&gt;
      &lt;/DataCollector&gt;
    &lt;/DataCollectors&gt;
  &lt;/DataCollectionRunSettings&gt;
&lt;/RunSettings&gt;

**Comment by AbhitejJohn** [comment]
@RomanKernSW : This works as well at my end. Are you performing an `Analyze Code Coverage`? Do you see a coverage file generated in the `TestResults` folder(Beside the solution file)?

**Comment by RomanKernSW** [comment]
Sorry, yes: TestResults is always generated. Code Coverage File is generated too, but it is empty except the first x bytes

**Comment by AbhitejJohn** [comment]
ah I see. It looks like there is an issue in collecting coverage data then. Was this working earlier? Can you share the coverage file that got generated please?

Tagging @mayankbansal018 here as well.

**Comment by RomanKernSW** [comment]
I've tested already the vs 2015 and 2017 calls. vstest.console in 2017 do generate ".coverage" file without any information in it. It just filled with some bytes. give the testadapter path not solving the problem (sure, it just for tests not for coverage). the vs 2015 version creates ".coverage" file with code coverage results for all tests.
Microsoft.CodeCoverage nuget is not helping too. thought, maybe the asp.netv5 instrumentations would help, but they dont.

codecoverage.zip is currrent runsetting config the other file is the code coverage output
[CodeCoverage.zip](https://github.com/Microsoft/testfx/files/957465/CodeCoverage.zip)

[2017-04-26 08_14_28.zip](https://github.com/Microsoft/testfx/files/957410/2017-04-26.08_14_28.zip)

**Comment by AbhitejJohn** [comment]
@RomanKernSW : Spoke with @mayankbansal018 offline. Can you flip `UseVerifiableInstrumentation` to false in the runsettings above. Also can you verify if you can get Code Coverage on a blank unit test project?
Is this an ASP.NET based project? The issue filed seems to be pointing to a .Net 4.5.2 runtime. A sample project where this repro's(if possible) would help as well.

**Comment by RomanKernSW** [comment]
@AbhitejJohn to the settings: Yes both are correct. MS Test V2 Code Coverage works with VS 2015 and 2017 not. Test are executed. There is no changes with VS 2017 if testadapterpath variable is set or not. Same output. 
Setting UseVerifiableInstrumentation to false emit no changes on output. I do still have the same scenario. No code coverage. 
dotnet-test-mstest nuget do not create positive outcome. still have no code coverage

**Comment by AbhitejJohn** [comment]
@RomanKernSW : Thanks for that info. How about the below:
>Also can you verify if you can get Code Coverage on a blank unit test project?
Is this an ASP.NET based project? The issue filed seems to be pointing to a .Net 4.5.2 runtime. A sample project where this repro's(if possible) would help as well.

VS 2017 workflows do not need `dotnet-test-mstest` nor do they need `Microsoft.CodeCoverage` nuget package. A smaller sized repro (if possible) would greatly help get to the bottom of this.

**Comment by RomanKernSW** [comment]
dotnet-test-mstest was not installed as default. I just thought, it can be usefull to try with this packages too. So I've tested with this packages once and uninstall them afterward.

creating new project do not resolve the problem. the framework itself seems to be irrelevant (tried .Net 4.6.1 too).

Besides that, I do have another problem(s). the variable testadapterpath do not accept quoted string. my new project contains spaces in the path. 

Error: The path 'C:\Users\xyz\Documents\visual studio 2017\Projects\MsTestV2
Testapp\MsTestV2Testapp.Test\bin\Debug" /UseVsixExtensions:true' specified in th
e 'TestAdapterPath' is invalid. Error: Illegales Zeichen im Pfad.
>means => illegal character in path

My VS 2015 is english and VS 2017 is german. this was english version. Window is german

if the testadapter is not in same folder as test assembly:
Error: An exception occurred while invoking executor 'executor://mstestadapter/v
2': Der Konstruktor für den Typ "Microsoft.VisualStudio.TestPlatform.MSTest.Test
Adapter.Execution.UnitTestRunner" wurde nicht gefunden.

>mean => no constructor found ... 

maybe its a correct behave in such situation.

**Comment by AbhitejJohn** [comment]
@RomanKernSW : Thanks again. 

>My VS 2015 is english and VS 2017 is german. this was english version. Window is german

That's interesting. So VS 2017 is German. There might be an issue in Code Coverage here. Can you also confirm if this is a 32 bit machine or a 64 bit machine?

**Comment by RomanKernSW** [comment]
Currently on Windows 7 64 Bit. I do have an windows 10 64 Bit with VS 2017. 

There is asp.net Core project, that i maintain, on visual studio online where CI run on hosted VS 2017 machine. There is no code coverage too. I dont know, if the hosted machine is 32 or 64 bit. MS Build achritecture x86 on this ci. 

The .net core is not running with x64 build, but the small unit test project with .net 4.5.2 can be started with x64. Code coverage is not available anyway.

**Comment by AbhitejJohn** [comment]
Just gave it a shot on both a 32 bit and a 64 bit German box with Vs 2017 German installed and this worked for me. Can you try it out on this project please: [ReproCodeCoverage.zip](https://github.com/Microsoft/testfx/files/970344/ReproCodeCoverage.zip)

**Comment by RomanKernSW** [comment]
haha... OK, your project still not working well for me. cant see the code coverage on 2017th.

hm.. but I still have same issue, because my other machine is Windows 10. Maybe I need to reinstall VS 2017. You have installed your VS 2017 on new Windows Image today or the other day without any updates?

very nice bug.  VS 2017 info:
Microsoft Visual Studio Enterprise 2017 
Version 15.1 (26403.7) Release
VisualStudio.15.Release/15.1.0+26403.7
Microsoft .NET Framework
Version 4.6.01590

**Comment by AbhitejJohn** [comment]
Yes this is a fresh install and on the very same build you have. Very interesting issue indeed. Is there a possibility you can give VS 2017 enu a try?

@mayankbansal018 anything else we can do to figure out what's happening here?

**Comment by RomanKernSW** [comment]
Hi, enu is acronym for what exactly?

**Comment by mayankbansal018** [comment]
Hi @RomanKernSW,

enu- is acronym for English.

On German VS can you please do AnalyzeCodeCoverage again, & check in Event Viewer for .net runtime message. I'm expecting something of this sort ".NET Runtime version 4.0.30319.0 - The profiler was loaded successfully.  Profiler CLSID: '{9317ae81-bcd8-47b7-aaa1-a28062e41c71}'.  Process ID (decimal): 20040.  Message ID: [0x2507]."

Please share the coverage file that gets generated for you in this process. So far this most likely seems to be installation/upgrade gone wrong.

Please also check if there is any Environment variable with name "VS150COMNTOOLS", if it is there please note it's value, & remove it. Make sure you restart VS after this.

**Comment by RomanKernSW** [comment]
>echo %VS150COMNTOOLS%
C:\Program Files (x86)\Microsoft Visual Studio 15.0\Common7\Tools\

>.NET Runtime version 4.0.30319.0 - Der Profiler wurde erfolgreich geladen. Profiler-CLSID: "{9317ae81-bcd8-47b7-aaa1-a28062e41c71}".  Prozess-ID (dezimal): 17012. Meldungs-ID: [0x2507].

[PCName 2017-05-03 11_44_42.zip](https://github.com/Microsoft/testfx/files/972944/PCName.2017-05-03.11_44_42.zip)

sometimes i have to repaire installation after update...

**Comment by AbhitejJohn** [comment]
@mayankbansal018 : Looked at the coverage file @RomanKernSW shared above and this one is empty too. 
```
<?xml version="1.0" encoding="UTF-8" ?>
<results>
</results>
```
and it looks like the TraceLog profiler seems to have been initialized without any problems. 

>I will try to install new VM with VS 2017 includes newest updates and try mstest v2 again.

@RomanKernSW : Just checking if this gave you similar results.

**Comment by RomanKernSW** [comment]
But what exactly cause this problem? VS installation/update problem or the defect is on nuget package (ms test vs2 assemblies)?

**Comment by AbhitejJohn** [comment]
This actually isn't related to the mstest v2 package since this works with the same package in VS 2015 and at our end with VS 2017 as well. Code coverage isn't also related to what framework is chosen. This as of now looks like an installation issue but I wouldn't jump to that conclusion just yet. I'l let @mayankbansal018 suggest what we can do next.

**Comment by RomanKernSW** [comment]
OK. So I have to forward this bug to VS Installation routine team?

**Comment by AbhitejJohn** [comment]
Can you send a mail to vstestsup@microsoft.com please? We can set up a call to figure this out.

**Comment by RomanKernSW** [comment]
Should I close this issue?

**Comment by AbhitejJohn** [comment]
Sure @RomanKernSW . Go ahead.

**Comment by RomanKernSW** [comment]
So its confirmed. my VS 2017 installation is broken. Other colleagues can see the code coverage without problem, one with fresh installed VS and another that installed all updates one after another.

**Comment by patsissons** [comment]
Ran into this issue today. I had installed 2017 professional first, then enterprise second, then uninstalled professional. Today i uninstalled enterprise (no remaining 2017 products installed), rebooted, and reinstalled enterprise to fix the issue.

**Comment by PascalN2** [comment]
We were also able to fix the problem by uninstalling VS Pro 2017, rebooting and then installing VS Enterprise 2016.

**Comment by RoseTharailJohn** [comment]
I have been having the same issue with the latest of VS Enterprise 2017 15.6.2. I tried uninstalling reinstalling it on my my build machine. The issue persists. The trx files created always show 0 coverage.

**Comment by halovanic** [comment]
I have the same results as @RoseTharailJohn.  I found that the coverage comes out as 0 when the vstest.console.exe is run under the _Local System_ account via a stock install of Jenkins.  Running under another account like my own or _Network Service_ made the coverage show up again.  It still shows some odd logging behavior with that Coverage Collection Tool message printing at the beginning **and** in the middle of the test execution for some reason.

I was able to reproduce this with both 15.5.7 and 15.6.4.  2015's vstest seems to work perfectly fine under the Local System account.
